### PR TITLE
[6562] Nowa zakładka - average hours by month

### PR DESCRIPTION
**Pull request by wiktorcie**
https://bugzilla.stxnext.pl/show_bug.cgi?id=6562

### CODE REVIEW

**Comment by mjelonek92** [comment]
The code is repeating from the previous PR. Please fix that, it shouldn't happen.

File: src/presence_analyzer/views.py [hunk 1]
```diff
@@ -102,6 +103,25 @@ def presence_start_end_view(user_id):
...
+    Returns average hours worked per month.
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    result = [
+        [calendar.month_abbr[i + 1], part]
+        for i, part in enumerate(group_by_average_monthly_hours(data[user_id]))
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: What is `i` and `part` in this context? Please use more descriptive names.

File: src/presence_analyzer/utils.py [hunk 2]
```diff
@@ -138,6 +171,37 @@ def group_by_average_start_end_time(items):
...
+    """
+    Groups average hours worked per month.
+    """
+    years = []
+    for year in items:
+        years.append(year.year)
+
+    result = {year: [[] for i in xrange(12)] for year in set(years)}
+
+    for day in items:
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: What's `items`?

File: src/presence_analyzer/utils.py [hunk 3]
```diff
@@ -138,6 +171,37 @@ def group_by_average_start_end_time(items):
     return result
 
 
+def group_by_average_monthly_hours(items):
+    """
+    Groups average hours worked per month.
+    """
+    years = []
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Use set comprehension.

File: src/presence_analyzer/utils.py [hunk 4]
```diff
@@ -138,6 +171,37 @@ def group_by_average_start_end_time(items):
...
+    for year in items:
+        years.append(year.year)
+
+    result = {year: [[] for i in xrange(12)] for year in set(years)}
+
+    for day in items:
+        start = seconds_since_midnight(items[day]['start'])
+        end = seconds_since_midnight(items[day]['end'])
+        hours_per_day = (end - start)/60/60
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Spaces around operators.

File: src/presence_analyzer/utils.py [hunk 5]
```diff
@@ -138,6 +171,37 @@ def group_by_average_start_end_time(items):
...
+
+    for year in result:
+        result[year] = [sum(month) for month in result[year]]
+
+    final = [[] for i in xrange(12)]
+
+    for year in result:
+        for month in xrange(12):
+            if result[year][month]:
+                final[month].append(result[year][month])
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Why do you group by year first if in the end you'll group by month?

File: src/presence_analyzer/utils.py [hunk 6]
```diff
@@ -138,6 +171,37 @@ def group_by_average_start_end_time(items):
...
+
+    final = [sum(x)/len(x) for x in final]
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: `x`?

File: src/presence_analyzer/utils.py (hunk 5 above)
Comment (wiktorcie) [inline src/presence_analyzer/utils.py]: It's necessary to correctly calculate average hours per month.
For example 
2011 have [140,156,130...]
2012 have [0, 0, 145....]

"if result[year][month]: " checks for months with no hours, where employee wasn't working , so we don't include them in average

result is [140, 156, 137,5....]

Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Oh, I see. Fair enough

Comment (wiktorcie) [inline src/presence_analyzer/utils.py]: `In [26]: years_with_all_months = {year: [[] for i in xrange(12)] for year in years}

In [27]: years_with_all_months
Out[27]: 
{**2011**: [[], [], [], [], [], [], [], [], [], [], [], []],
 **2012**: [[], [], [], [], [], [], [], [], [], [], [], []],
 **2013**: [[], [], [], [], [], [], [], [], [], [], [], []]}
`
You mean bolded this part?

Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: I don't think you need years at all for what you're trying to achieve. Just group everything by month, filter out zeroes and calculate average?

File: src/presence_analyzer/utils.py [hunk 7]
```diff
@@ -138,6 +171,36 @@ def group_by_average_start_end_time(items):
...
+    """
+    years = {year.year for year in user_data}
+
+    years_with_all_months = {year: [[] for i in xrange(12)] for year in years}
+
+    for day in user_data:
+        start = seconds_since_midnight(user_data[day]['start'])
+        end = seconds_since_midnight(user_data[day]['end'])
+        hours_per_day = (end - start) / 60 / 60
+        years_with_all_months[day.year][day.month - 1].append(hours_per_day)
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Why not group here by month already? You're doing that in the end. You don't need this year information if you're not going to use it in the end. List that you're returning (`final`) has no information about the year whatsoever.

**Comment by mjelonek92** [comment]
Missing tests.

File: src/presence_analyzer/views.py [hunk 8]
```diff
@@ -102,6 +103,26 @@ def presence_start_end_view(user_id):
...
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    result = [
+        [calendar.month_abbr[month_count + 1], number_of_hours]
+        for month_count, number_of_hours
+        in enumerate(group_by_average_monthly_hours(data[user_id]))
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: The second argument to `enumerate` is `start` which means that you don't need to do `+ 1` if you specify `start=1`.

File: src/presence_analyzer/utils.py (hunk 7 above)
Comment (wiktorcie) [inline src/presence_analyzer/utils.py]: You mean to append all this individual hours so I end up with something like [523, 425, 622, 425, 622, 425, 622...] ?

I won't have a way to know how many months the hours span.

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -128,6 +128,32 @@ def test_api_average_start_end_time(self):
         self.assertEqual(resp.content_type, 'application/json')
         self.assertEqual(expected_data, received_data)
 
+    def test_api_average_by_month(self):
+        """
+        Tests average hourly presence by month api respone.
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: typo: `respone` -> `response`

File: src/presence_analyzer/utils.py [hunk 10]
```diff
@@ -171,6 +172,29 @@ def group_by_average_start_end_time(items):
...
+        start = seconds_since_midnight(times['start'])
+        end = seconds_since_midnight(times['end'])
+        hours_per_day = (end - start) / 60 / 60
+        months[day.month].append(hours_per_day)
+
+    result = [[] for i in xrange(12)]
+
+    for month, hours in months.iteritems():
+        result[month - 1] = (
+            sum(hours) / len({(day.year, day.month) for day in user_data
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Could you calculate the `len` in a separate statement? It got really cluttered here.

**Comment by mjelonek92** [comment]
:+1:

**Pull request merged**